Addon-docs/Angular: Add dynamic source snippets

yngvebn opened the pull request:
Issue: #13555

## Todo

- [x] Enable proper source for templated stories (see remarks)
- [x] Get the build passing
- [ ] Source doesn't update with iframe-rendering

## What I did
This PR implements the dynamic snippet feature with Angular.

~Currently it only supports argsStory, but it should be possible to make some "lightweight" support of other storytypes as well.~

**Remarks**
https://github.com/storybookjs/storybook/pull/13525 fixes both of the issues mentioned here
 ~- Currently the "Show source" button doesn't enable until you either refresh or change one of the args in the ArgsTable. Would be great if someone could shed some light on this~
 ~- Unfortunately the source will not update when Args are changed. This will most likely be resolved as part of this PR https://github.com/storybookjs/storybook/pull/13525, but I have not verified this.~

How this works:

~If the story is an ArgsStory~ the source-code will be automatically generated based on Component-annotations and props.

#### ArgsStory
Due to this check in `/addon/docs/src/blocks/Source.tsx` 
```  if (type === SourceType.AUTO && snippet && isArgsStory) {```

This will currently only display source for ArgsStory. I'm not touching this now, as it's not clear for me the implications this might have on other parts of Storybook

*UPDATE: Set default type to SourceType.DYNAMIC for Angular in docs-config fixed the abovementioned problem*

## How to test

- Is this testable with Jest or Chromatic screenshots? Done
- Does this need a new example in the kitchen sink apps? Done
- Does this need an update to the documentation? I don't think

If your answer is yes to any of these, please make sure to include it in your PR.

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this.

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

yngvebn committed:
Angular: implement dynamic snippet

ThibaudAV reviewed (commented):
Hello 👋 , 

Thank you for your contribution :). 
I have some questions

I find it strange that the sources are generated from `args` and not from story `props`

More exactly, I would expect that the template in "show source" would be the same as the one used in story render. 

if you agree with this. I think it's possible to do it. And it would also allow to simplify and mutualize some code 🤔  

do you think angular output should be displayed on source as well ?

yngvebn commented:
> Hello 👋 ,
> 
> Thank you for your contribution :).
> I have some questions
> 
> I find it strange that the sources are generated from `args` and not from story `props`
> 
> More exactly, I would expect that the template in "show source" would be the same as the one used in story render.
> 
> if you agree with this. I think it's possible to do it. And it would also allow to simplify and mutualize some code 🤔
> 
> do you think angular output should be displayed on source as well ?

Changed this to use props instead of args, and moved the source-logic into the angular-beta/ComputesTemplateFromComponent file for consolidation of code. :)

yngvebn commented:
I don't understand why the checks are failing. Can someone help me look at it?

shilman commented:
@yngvebn I've updated the snapshots on my dev machine and CI is passing now. Are you developing on Windows?

There are still a few chromatic rendering errors. I'll look into those later today.

yngvebn commented:
> @yngvebn I've updated the snapshots on my dev machine and CI is passing now. Are you developing on Windows?
> 
> There are still a few chromatic rendering errors. I'll look into those later today.

Yeah, I'm developing on windows, and the only errors I'm getting locally now are related to filepaths, so I'm assuming that's a windows issue.

ThibaudAV commented on addons/docs/src/frameworks/angular/sourceDecorator.ts:
```diff
@@ -0,0 +1,57 @@
+ * @param storyFn Fn
+ * @param context  StoryContext
+ */
+export const sourceDecorator = (storyFn: any, context: StoryContext) => {
```
```suggestion
export const sourceDecorator = (storyFn: StoryFn<IStory>, context: StoryContext) => {
```

`import { IStory } from '@storybook/angular';`

ThibaudAV commented on app/angular/src/client/preview/angular-beta/ComputesTemplateFromComponent.test.ts:
```diff
@@ -0,0 +1,162 @@
+import { ButtonAccent, InputComponent, ISomeInterface } from './__testfixtures__/input.component';
+
+describe('angular source decorator', () => {
+  it('With no args should generate simple tag', () => {
```
now is `props` can we change the tests so as not to cause confusion?

ThibaudAV commented on app/angular/src/client/preview/angular-beta/ComputesTemplateFromComponent.ts:
```diff
@@ -52,3 +53,66 @@ export const computesTemplateFromComponent = (
+}: {
+  propertyName: string;
+  value: any;
+  argType: ArgType;
```
```suggestion
  argType?: ArgType;
```

could be undefined si `ArgTypes` is undefined to, no ?

ThibaudAV commented on app/angular/src/client/preview/angular-beta/ComputesTemplateFromComponent.ts:
```diff
@@ -52,3 +53,66 @@ export const computesTemplateFromComponent = (
+  const requiresQuotes =
+    (type === 'enum' && !summary) || ['string', 'any', 'object'].includes(type || typeof value);
+
+  if (category === 'outputs' || typeof value === 'function') {
```
only angular input is managed here no ? 
output is with templateOutputs

ThibaudAV commented on app/angular/src/client/preview/angular-beta/__testfixtures__/input.component.ts:
```diff
@@ -0,0 +1,187 @@
+// @ts-nocheck
+/* eslint-disable no-console */
```
as we are in tests it may be preferable to have no any console.log ?

ThibaudAV commented on app/angular/src/client/preview/angular-beta/ComputesTemplateFromComponent.ts:
```diff
@@ -52,3 +53,66 @@ export const computesTemplateFromComponent = (
+};
+
+/**
+ * Converted a component into a template with inputs/outputs present in initial props
+ * @param component
+ * @param initialProps
+ * @param innerTemplate
```
```suggestion
 * Converted a component into a displayable template source with inputs/outputs present in initial props
 * @param component
 * @param initialProps
 * @param argTypes
```

ThibaudAV commented on app/angular/src/client/preview/angular-beta/__testfixtures__/input.component.ts:
```diff
@@ -0,0 +1,187 @@
+  selector: 'doc-button',
+  template: '<button>{{ label }}</button>',
+})
+export class InputComponent<T> {
```
I think we're missing a test with 
```
  @Input('inputBindingPropertyName')
  public localPropertyName: string;
```
to complete the picture :) 
with your first version of code, it wouldn't have worked. now, I think it works without any problem

ThibaudAV commented on app/angular/src/client/preview/angular-beta/__testfixtures__/input.component.ts:
```diff
@@ -0,0 +1,187 @@
+   *
+   * @ignore
+   */
+  public get classes(): string[] {
```
Can we simplify this test component? Some properties are not necessary for tests like this one, no?

ThibaudAV reviewed (commented):
I can't understand when the story source is displayed and when it's not. 
nor why even when a `SNIPPET_RENDERED` is emitted it does not display it. 🤷‍♂️ 

Sad, netlify build didn't work we could have seen the different cases :/ 


what do you think to add 
```
<Source story="with text" />
```
in simple.stories.mdx
  


💪   nice work :) I was also looking forward to this feature on docs 🙏 

I have a few comments and adjustments to make the code as pretty as possible

ThibaudAV reviewed (commented):
nice, LGTM :) 

i see some test fail because you are a windows user :p shame on you 
no its a joke 😁 

I don't know if it's your IDE, git, terminal or another problem. but something is misconfigured 
normally the `end_of_line = lf` of the [editorconfig](https://github.com/storybookjs/storybook/blob/next/.editorconfig#L4) should solve this kind of problem (fight between `\r\n` or `\n`)

ThibaudAV commented:
The unit tests are green. :) 

but the [netlify build failed](https://app.netlify.com/sites/storybookjs/deploys/6019d319f672a200072be5a3). I think, for a reason unrelated to this PR. sad

yngvebn commented:
> The unit tests are green. :)
> 
> but the [netlify build failed](https://app.netlify.com/sites/storybookjs/deploys/6019d319f672a200072be5a3). I think, for a reason unrelated to this PR. sad

Excellent, thanks! I did find that VSCode didn't honor the .editorconfig end-of-line setting, but you can override it in settings. That doesn't change the fact that `yarn test --update` still generates with /r/n, but now that I know that's the issue, I'll see if I can figure out how to fix it, and issue a PR on CONTRIBUTING.md with instructions :)

shilman reviewed (commented):
This is looking great!

I kicked the tires and saw a few bits that could be improved. This is in `examples/angular-cli` in the monorepo:

<img width="806" alt="Addon___Controls_-_Basic_⋅_Storybook" src="https://user-images.githubusercontent.com/488689/106711481-8c1c6080-6632-11eb-86a6-136ddd0cf141.png">

- Need to remove an extra layer of quotes (I think)
- Need to JSON stringify objects

ThibaudAV commented:
> Need to remove an extra layer of quotes (I think)
We talked about it a little bit. I agree with you @shilman . But I don't have a real argument. (except that it is shorter) so 🤷‍♂️. maybe we should ask for feedback
it would give a result like this  ? :
```
<my-button
  appearance="secondary"
  [isDisabled]="false"
  [someDataObject]="'{ test: "test" }'"
  label="Args test"
  size="medium"
></my-button>
```

shilman commented:
I'll defer to you guys on the quote issue. I'm not an Angular user, so I don't know what's customary there. I do think that the `JSON.stringify` needs fixing, however. Probably the simplest thing would be to simply escape the double quotes in the output using something like this: https://gist.github.com/getify/3667624

yngvebn commented:
> I'll defer to you guys on the quote issue. I'm not an Angular user, so I don't know what's customary there. I do think that the `JSON.stringify` needs fixing, however. Probably the simplest thing would be to simply escape the double quotes in the output using something like this: https://gist.github.com/getify/3667624

The problem I see with JSON.stringifying complex objects is that often you will have an input that takes either a fairly large object, or an array of such. This will probably make the source a lot less readable, and is most likely not what you're going to want to use when you copy the source to paste it into your own code. The way it's solved in the latest changes is that an object is simple rendered as `[someDataObject]="someDataObject"` which will indicate that you will have to have that object as a property on your component. I'm happy to hear others thoughts as well, though.

shilman reviewed (approved):
Looking great. If @ThibaudAV is good with it, I'm good with it.

ThibaudAV reviewed (approved):
🚀  LGTM.  

It's already much better than it was before xD . The code will allow an improvement if needed. And , it is linked to native operation of renderer which will make it, I hope, easily maintainable.
